File: src/core/passes/flow-x-reuse.ts
import type { IrOp, RegisterName } from "../types.ts";
import { isStableIndirectSelector } from "../indirect-addressing.ts";
import { buildCfgEdges, buildTargetIndexes, loopCounterRegister, type CfgEdge } from "./cfg.ts";
import {
  hasRewriteBarrier,
  knownIndirectFlowTarget,
  knownIndirectMemoryTarget,
  plainPreservesXValue,
  removableRecallValueRegister,
  storedCurrentXValueRegister,
  type IrPass,
  type IrPassFn,
} from "./helpers.ts";
import { runRecallRemovalPass } from "./recall-removal.ts";

type XRegisterSet = ReadonlySet<RegisterName>;

const run: IrPassFn = (ops) =>
  runRecallRemovalPass(
    ops,
    {
      name: "flow-x-reuse",
      detail: (count) =>
        `Dropped ${count} recall${count === 1 ? "" : "s"} whose register value already reaches the point in X on every CFG predecessor.`,
    },
    (engine) => {
      if (ops.length === 0) return;
      if (hasUnknownIndirectFlow(ops)) return;
      const numericTargets = numericFlowTargetLayoutGuard(ops);
      if (numericTargets === undefined) return;

      const graph = buildCfgEdges(ops);
      const inStates = computeXRegisterStates(ops, graph);

      for (let index = 0; index < ops.length; index += 1) {
        const op = ops[index]!;
        const recallRegister = removableRecallValueRegister(op);
        if (recallRegister === undefined) continue;
        if (!numericTargets.canRemoveAt(index)) continue;
        const removalPlan = engine.plan(index);
        if (removalPlan?.removable !== true) continue;
        const alreadyInX =
          inStates[index]?.has(recallRegister) === true ||
          removalPlan.analysis.valueProof?.inX === true;
        if (alreadyInX) engine.removed.add(index);
      }
    },
  );

export const flowXReuse: IrPass = {
  name: "flow-x-reuse",
  run,
  layoutSafe: false,
};

function computeXRegisterStates(ops: readonly IrOp[], graph: CfgEdge[][]): Array<XRegisterSet | undefined> {
  const inStates: Array<Set<RegisterName> | undefined> = Array.from({ length: ops.length }, () => undefined);
  const outStates: Array<Set<RegisterName> | undefined> = Array.from({ length: ops.length }, () => undefined);
  inStates[0] = new Set();

  let changed = true;
  let iterations = 0;
  while (changed && iterations < 200) {
    changed = false;
    iterations += 1;

    for (let index = 0; index < ops.length; index += 1) {
      const input = inStates[index];
      if (input === undefined) continue;

      const output = transferXSet(input, ops[index]!, "normal");
      if (!sameSet(output, outStates[index])) {
        outStates[index] = output;
        changed = true;
      }

      for (const edge of graph[index] ?? []) {
        const edgeOutput = edge.kind === "normal" ? output : transferXSet(input, ops[index]!, edge.kind);
        const joined = joinXSets(inStates[edge.target], edgeOutput);
        if (!sameSet(joined, inStates[edge.target])) {
          inStates[edge.target] = joined;
          changed = true;
        }
      }
    }
  }

  return inStates;
}

function transferXSet(input: XRegisterSet, op: IrOp, edge: CfgEdge["kind"]): Set<RegisterName> {
  if (hasRewriteBarrier(op)) return new Set();

  switch (op.kind) {
    case "label":
    case "jump":
    case "cjump":
    case "orphan-address":
      return new Set(input);
    case "store":
    case "indirect-store": {
      const target = storedCurrentXValueRegister(op);
      return target === undefined ? new Set() : addRegister(input, target);
    }
    case "recall":
      return new Set([op.register]);
    case "indirect-recall": {
      const target = knownIndirectMemoryTarget(op);
      return target === undefined ? new Set() : new Set([target]);
    }
    case "plain":
      return plainPreservesXValue(op) ? new Set(input) : new Set();
    case "stop":
      return new Set();
    case "loop":
      return removeRegister(input, loopCounterRegister(op.counter));
    case "call":
    case "return":
      return new Set(input);
    case "indirect-jump":
    case "indirect-call":
      return transferIndirectFlowXSet(input, op.register);
    case "indirect-cjump":
      return edge === "jump" ? transferIndirectFlowXSet(input, op.register) : new Set(input);
  }
}

function transferIndirectFlowXSet(input: XRegisterSet, register: RegisterName): Set<RegisterName> {
  const output = new Set(input);
  if (!isStableIndirectSelector(register)) output.delete(register);
  return output;
}

function addRegister(input: XRegisterSet, register: RegisterName): Set<RegisterName> {
  const output = new Set(input);
  output.add(register);
  return output;
}

function removeRegister(input: XRegisterSet, register: RegisterName): Set<RegisterName> {
  const output = new Set(input);
  output.delete(register);
  return output;
}

function joinXSets(current: Set<RegisterName> | undefined, incoming: XRegisterSet): Set<RegisterName> {
  if (current === undefined) return new Set(incoming);
  const joined = new Set<RegisterName>();
  for (const register of current) {
    if (incoming.has(register)) joined.add(register);
  }
  return joined;
}

function sameSet(left: XRegisterSet | undefined, right: XRegisterSet | undefined): boolean {
  if (left === undefined || right === undefined) return left === right;
  if (left.size !== right.size) return false;
  for (const value of left) {
    if (!right.has(value)) return false;
  }
  return true;
}

interface NumericFlowTargetLayoutGuard {
  canRemoveAt(index: number): boolean;
}

function numericFlowTargetLayoutGuard(ops: readonly IrOp[]): NumericFlowTargetLayoutGuard | undefined {
  const { addressIndex } = buildTargetIndexes(ops);
  let latestTargetIndex = -1;
  for (const op of ops) {
    const target = numericFlowTarget(op);
    if (target === undefined) continue;
    const targetIndex = addressIndex.get(target);
    if (targetIndex === undefined) return undefined;
    latestTargetIndex = Math.max(latestTargetIndex, targetIndex);
  }
  return {
    canRemoveAt: (index) => index >= latestTargetIndex,
  };
}

function numericFlowTarget(op: IrOp): number | undefined {
  switch (op.kind) {
    case "jump":
    case "cjump":
    case "call":
    case "loop":
      return typeof op.target === "number" ? op.target : undefined;
    case "orphan-address":
      return typeof op.target === "number" ? op.target : undefined;
    default:
      return undefined;
  }
}

function hasUnknownIndirectFlow(ops: readonly IrOp[]): boolean {
  return ops.some((op) =>
    (op.kind === "indirect-jump" ||
      op.kind === "indirect-call" ||
      op.kind === "indirect-cjump") &&
    knownIndirectFlowTarget(op) === undefined
  );
}

File: native/src/core/passes/flow_x_reuse.cpp
#include "mkpro/core/passes/flow_x_reuse.hpp"

#include "mkpro/core/indirect_addressing.hpp"
#include "mkpro/core/passes/cfg.hpp"
#include "mkpro/core/passes/recall_removal.hpp"

#include <algorithm>
#include <deque>
#include <map>
#include <optional>
#include <set>
#include <string>
#include <utility>
#include <vector>

namespace mkpro::core::passes {

namespace {

using XRegisterSet = std::set<std::string>;

bool same_set_value(const XRegisterSet& left, const std::optional<XRegisterSet>& right) {
  return right.has_value() && left == *right;
}

XRegisterSet add_register(const XRegisterSet& input, const std::string& register_name) {
  XRegisterSet output = input;
  output.insert(register_name);
  return output;
}

XRegisterSet remove_register(const XRegisterSet& input, const std::string& register_name) {
  XRegisterSet output = input;
  output.erase(register_name);
  return output;
}

XRegisterSet join_x_sets(const std::optional<XRegisterSet>& current, const XRegisterSet& incoming) {
  if (!current.has_value())
    return incoming;
  XRegisterSet joined;
  for (const std::string& register_name : *current) {
    if (incoming.contains(register_name))
      joined.insert(register_name);
  }
  return joined;
}

XRegisterSet transfer_indirect_flow_x_set(const XRegisterSet& input,
                                          const std::string& register_name) {
  // "Stable" means no counter increment/decrement, not preservation of the
  // complete data word (e.g. 14.375 becomes 14). Without a word-level proof
  // only facts about other registers survive the selector operation.
  return remove_register(input, register_name);
}

XRegisterSet transfer_x_set(const XRegisterSet& input, const IrOp& op, CfgEdgeKind edge) {
  if (has_rewrite_barrier(op))
    return {};

  switch (op.kind) {
  case IrKind::Label:
  case IrKind::Jump:
  case IrKind::CondJump:
  case IrKind::OrphanAddress:
    return input;
  case IrKind::Store:
  case IrKind::IndirectStore:
    if (const std::optional<std::string> target = stored_current_x_value_register(op))
      return add_register(input, *target);
    return {};
  case IrKind::Recall:
    return XRegisterSet{op.register_name};
  case IrKind::IndirectRecall:
    if (const std::optional<std::string> target = known_indirect_memory_target(op))
      return XRegisterSet{*target};
    return {};
  case IrKind::Plain:
    return plain_preserves_x_value(op) ? input : XRegisterSet{};
  case IrKind::Stop:
    return {};
  case IrKind::Loop:
    return remove_register(input, loop_counter_register(op.counter));
  case IrKind::Call:
  case IrKind::Return:
    return input;
  case IrKind::IndirectJump:
  case IrKind::IndirectCall:
    return transfer_indirect_flow_x_set(input, op.register_name);
  case IrKind::IndirectCondJump:
    return edge == CfgEdgeKind::Jump ? transfer_indirect_flow_x_set(input, op.register_name)
                                     : input;
  }
  return {};
}

std::vector<std::optional<XRegisterSet>>
compute_x_register_states(const std::vector<IrOp>& ops,
                          const std::vector<std::vector<CfgEdge>>& graph) {
  std::vector<std::optional<XRegisterSet>> in_states(ops.size());
  std::deque<std::size_t> worklist;
  std::vector<bool> queued(ops.size(), false);
  if (!ops.empty())
    in_states.at(0) = XRegisterSet{};
  if (!ops.empty()) {
    worklist.push_back(0);
    queued.at(0) = true;
  }
  // Finite descending sets converge regardless of virtual program length or
  // block order. Never use partially propagated facts after an iteration cap.
  while (!worklist.empty()) {
    const std::size_t index = worklist.front();
    worklist.pop_front();
    queued.at(index) = false;
    const XRegisterSet input = *in_states.at(index);
    for (const CfgEdge& edge : graph.at(index)) {
      const std::size_t target = static_cast<std::size_t>(edge.target);
      const XRegisterSet joined =
          join_x_sets(in_states.at(target), transfer_x_set(input, ops.at(index), edge.kind));
      if (!same_set_value(joined, in_states.at(target))) {
        in_states.at(target) = joined;
        if (!queued.at(target)) {
          queued.at(target) = true;
          worklist.push_back(target);
        }
      }
    }
  }

  return in_states;
}

} // namespace

PassResult flow_x_reuse(const std::vector<IrOp>& ops, const PassContext& context) {
  (void)context;
  return run_recall_removal_pass(
      ops,
      RecallRemovalReport{
          .name = "flow-x-reuse",
          .detail =
              [](int count) {
                return "Dropped " + std::to_string(count) + " recall" + (count == 1 ? "" : "s") +
                       " whose register value already reaches the point in X on every CFG "
                       "predecessor.";
              },
      },
      [&](RecallRemovalEngine& engine) {
        if (ops.empty())
          return;
        const ControlFlowGraph control = build_control_flow_graph(ops);
        // Typed labels and finite target sets are exact even when no physical
        // byte can encode their current logical positions. Numeric-only
        // convenience queries must not veto an otherwise proved CFG.
        if (!control.targets_are_exact())
          return;

        const std::optional<NumericFlowTargetLayoutGuard> numeric_targets =
            numeric_flow_target_layout_guard(ops);
        if (!numeric_targets.has_value())
          return;

        const std::vector<std::optional<XRegisterSet>> in_states =
            compute_x_register_states(ops, control.edges);

        for (std::size_t index = 0; index < ops.size(); ++index) {
          const std::optional<std::string> recall_register =
              removable_recall_value_register(ops.at(index));
          if (!recall_register.has_value())
            continue;
          if (!numeric_targets->can_delete_at(static_cast<int>(index)))
            continue;

          const bool cfg_already_in_x =
              in_states.at(index).has_value() && in_states.at(index)->contains(*recall_register);
          RecallRemovalPlanOverrides overrides;
          overrides.require_value_proof = !cfg_already_in_x;
          const std::optional<RecallRemovalStackSchedulerPlan> removal_plan =
              engine.plan(static_cast<int>(index), overrides);
          if (!removal_plan.has_value() || !removal_plan->removable)
            continue;

          const bool value_proves_in_x = removal_plan->analysis.value_proof.has_value() &&
                                         removal_plan->analysis.value_proof->in_x;
          if (cfg_already_in_x || value_proves_in_x)
            engine.removed().insert(static_cast<int>(index));
        }
      });
}

IrPass flow_x_reuse_pass() {
  return IrPass{
      .name = "flow-x-reuse",
      .run = flow_x_reuse,
      .layout_safe = false,
  };
}

PassResult stack_lift_recall_forwarding(const std::vector<IrOp>& ops,
                                        const PassContext& context) {
  (void)context;
  const ControlFlowGraph control = build_control_flow_graph(ops);
  if (ops.empty() || !control.targets_are_exact())
    return PassResult{.ops = ops};
  const auto in_states = compute_x_register_states(ops, control.edges);
  RecallRemovalEngine engine(ops);
  std::vector<IrOp> result = ops;
  int forwarded = 0;
  for (std::size_t index = 0; index < ops.size(); ++index) {
    const IrOp& recall = ops[index];
    // Only a plain direct recall is replaced. Raw/manual cells and role-bound
    // opcode/data overlays require a separate byte-identity proof.
    if (recall.kind != IrKind::Recall || has_rewrite_barrier(recall) ||
        !recall.meta.roles.empty() || !in_states[index].has_value() ||
        !in_states[index]->contains(recall.register_name))
      continue;
    const auto plan = engine.plan(static_cast<int>(index));
    if (!plan.has_value() || !plan->analysis.exposes_stack_lift ||
        plan->analysis.exposes_x2_restore)
      continue;

    // X already equals the register on every predecessor. Enter duplicates
    // that same word instead of fetching it, preserving the required stack
    // lift. The shared recall proof establishes that X2 resynchronization is
    // unobserved. This one-cell rewrite cannot move even a fixed numeric
    // target, but can make the register's defining store dead in the next DSE.
    IrOp& replacement = result[index];
    replacement.kind = IrKind::Plain;
    replacement.opcode = 0x0e;
    replacement.register_name.clear();
    replacement.semantic.clear();
    replacement.meta.logical_register_name.reset();
    replacement.meta.mnemonic = "В↑";
    replacement.meta.comment = "current-X forwarding with preserved stack lift";
    replacement.meta.tactic = "stack-lift-recall-forwarding";
    ++forwarded;
  }
  if (forwarded == 0)
    return PassResult{.ops = ops};
  return PassResult{
      .ops = std::move(result),
      .applied = forwarded,
      .optimizations = {{
          .name = "stack-lift-recall-forwarding",
          .detail = "Replaced " + std::to_string(forwarded) +
                    " current-X recall(s) with Enter, preserving stack lifts and "
                    "physical positions while exposing dead register stores.",
      }},
  };
}

IrPass stack_lift_recall_forwarding_pass() {
  return IrPass{
      .name = "stack-lift-recall-forwarding",
      .run = stack_lift_recall_forwarding,
      .layout_safe = true,
  };
}

} // namespace mkpro::core::passes
